docs/homepage/blog/ospp_mid-term_report_210370190/index.md

```diff
@@ -0,0 +1,400 @@
+    mode::Bool # `true` for best response mode(RL agent's policy), `false` for  average policy mode(SL agent's policy). Only used in training.
+end
+```
+Based on 2.1, the core of the `NFSPAgent` is customized behaviors on the specific stage:
+
+- PreEpisodeStage
+
+Here, `NFSPAgent` should set train mode based on the anticipatory dynamics\dcite{1406126} and delete the terminated state and dummy action if having gone through one episode before. Note that here deleting the terminated state and dummy action is necessary for the algo(see the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134)), otherwise may occur to have some unreliable samples.
+```Julia
+function (π::NFSPAgent)(stage::PreEpisodeStage, env::AbstractEnv, ::Any)
+    # delete the terminal state and dummy action.
+    update!(π.rl_agent.trajectory, π.rl_agent.policy, env, stage)
+
+    # set the train's mode before the episode.(anticipatory dynamics)
+    π.mode = rand(π.rng) < π.η
+end
+```
+
+- PreActStage
+
+In this stage, `NFSPAgent` should collect the personal information `state` and `action` to the RL agent's trajectory, and if on the `best response mode`, also update the SL agent's trajectory. Besides, if satisfying the condition of updating, here also need to update the inner agents. The code is just like the following:
+```Julia
+function (π::NFSPAgent)(stage::PreActStage, env::AbstractEnv, action)
+    rl = π.rl_agent
+    sl = π.sl_agent
+    # update trajectory
+    if π.mode
+        update!(sl.trajectory, sl.policy, env, stage, action)
+        rl(stage, env, action)
+    else
+        update!(rl.trajectory, rl.policy, env, stage, action)
+    end
+
+    # update policy
+    π.update_step += 1
+    if π.update_step % π.update_freq == 0
+        if π.mode
+            update!(sl.policy, sl.trajectory)
+        else
+            rl_learn!(rl.policy, rl.trajectory)
+            update!(sl.policy, sl.trajectory)
+        end
+    end
+end
+```
+
+- PostActStage
+
+Here, the agent needs to collect the personal `reward` and  the `is_terminated` judgment of the current state to the RL agent's trajectory.
+```Julia
+function (π::NFSPAgent)(::PostActStage, env::AbstractEnv, player::Any)
+    push!(π.rl_agent.trajectory[:reward], reward(env, player))
+    push!(π.rl_agent.trajectory[:terminal], is_terminated(env))
+end
+```
+
+- PostEpisodeStage
+
+When one episode is terminated, the agent should collect the terminated state and a dummy action(see the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134)) to the RL agent's trajectory. Also, the reward and terminal judgment need to be corrected to avoid getting wrong samples when playing the sequential or terminal_reward games.
```

Comment:
```suggestion
When one episode is terminated, the agent should push the terminated state and a dummy action(see the [note](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/4e5d258798088b1c628401b6b9de18aa8cbb3ab3/src/ReinforcementLearningCore/src/policies/agents/agent.jl#L134)) into the **RL** agent's trajectory. Also, the reward and terminal judgment need to be corrected to avoid getting wrong samples when playing the sequential or terminal_reward games.
```